dali/imgcodec/decoders/nvjpeg2k.h

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_DECODERS_NVJPEG2K_H_
+#define DALI_IMGCODEC_DECODERS_NVJPEG2K_H_
+
+#include <nvjpeg.h>
+#include <memory>
+#include <vector>
+#include "dali/imgcodec/decoders/decoder_parallel_impl.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg2k_helper.h"
+#include "dali/imgcodec/decoders/nvjpeg/nvjpeg_memory.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+namespace imgcodec {
+
+/**
+ * @brief Decoder interface for nvjpeg2k library
+ */
+class DLL_PUBLIC NvJpeg2000DecoderInstance : public BatchParallelDecoderImpl {
+ public:
+  NvJpeg2000DecoderInstance(int device_id, ThreadPool *tp);
+
+  using BatchParallelDecoderImpl::CanDecode;
+  bool CanDecode(ImageSource *in, DecodeParams opts, const ROI &roi) {
+    return !roi && (opts.dtype == DALI_UINT8);
+  }
+
+  using BatchParallelDecoderImpl::DecodeImplTask;
+  DecodeResult DecodeImplTask(int thread_idx,
+                              cudaStream_t stream,
+                              SampleView<GPUBackend> out,
+                              ImageSource *in,
+                              DecodeParams opts,
+                              const ROI &roi) override;
+
+  void SetParam(const char *name, const any &value) {
+    if (strcmp(name, "nvjpeg2k_device_memory_padding") == 0) {
+      nvjpeg2k_device_memory_padding_ = any_cast<size_t>(value);
+    } else if (strcmp(name, "nvjpeg2k_host_memory_padding") == 0) {
+      nvjpeg2k_host_memory_padding_ = any_cast<size_t>(value);
+    }
```

Comment:
```suggestion
    } else {
      DALI_FAIL("Unrecognized param name");
    }
```

Reply:
I think that SetParam shouldn't panic when receiving unrecognized params. This way they can be easily populated when holding some unspecified `ImageDecoderInstance` at the higher level.

Reply:
That's a very good point. Thanks!